src/main/java/htsjdk/samtools/reference/FastaReferenceWriter.java

```diff
@@ -0,0 +1,758 @@
+     * </p>
+     * <p>
+     * The description cannot contain {@link Character#isISOControl(char)}. If set to {@code null} or the empty
+     * string ("") no description will be outputted.
+     * </p>
+     * <p>
+     * The input bases-per-line must be 1 or greater.
+     * </p>
+     * <p>
+     * This method cannot be called after the writer has been closed.
+     * </p>
+     * <p>
+     * It also will fail if no base was added to the previous sequence if any.
+     * </p>
+     *
+     * @param sequenceName the name of the new sequence.
+     * @param description  optional description for that sequence.
+     * @param basesPerLine number of bases per line for this sequence.
+     * @return this instance.
+     * @throws IllegalArgumentException if any argument does not comply with requirements listed above.
+     * @throws IllegalStateException    if no base was added to the previous sequence or the writer is already closed of
+     *                                  the sequence has been already added.
+     * @throws IOException              if such exception is thrown when writing into the output resources.
+     */
+    public FastaReferenceWriter startSequence(final String sequenceName, final String description, final int basesPerLine)
+            throws IOException {
+        assertIsNotClosed();
+        checkSequenceName(sequenceName);
+        final String nonNullDescription = checkDescription(description);
+        checkBasesPerLine(basesPerLine);
+        closeSequence();
+        if (sequenceNamesAndSizes.containsKey(sequenceName)) {
+            throw new IllegalStateException("the input sequence name '" + sequenceName + "' has already been added");
+        }
+        currentSequenceName = sequenceName;
+        currentBasesPerLine = basesPerLine;
+        final StringBuilder builder = new StringBuilder(sequenceName.length() + nonNullDescription.length() + 10);
+        builder.append(HEADER_START_CHAR).append(sequenceName);
+        if (!nonNullDescription.isEmpty()) {
+            builder.append(HEADER_NAME_AND_DESCRIPTION_SEPARATOR).append(nonNullDescription);
+        }
+        fastaStream.write(builder.toString().getBytes(CHARSET));
+        fastaStream.write(LINE_SEPARATOR);
+        currentSequenceOffset = fastaStream.getBytesWritten();
+        return this;
+    }
+
+    private void closeSequence()
+            throws IOException {
+        if (currentSequenceName != null) {
+            if (currentBasesCount == 0) {
+                throw new IllegalStateException("no base was added");
+            }
+            sequenceNamesAndSizes.put(currentSequenceName, currentBasesCount);
+            writeIndexEntry();
+            writeDictEntry();
+            fastaStream.write(LINE_SEPARATOR);
+            currentBasesCount = 0;
+            currentLineBasesCount = 0;
+            currentSequenceName = null;
+        }
+    }
+
+    private void writeIndexEntry()
```

Comment:
Isn't this functionality somewhere already? Otherwise, it will be nice to have an index-codec for decoding/encoding

Reply:
Or a `FastaIndexing` class to index on the fligh streams

Reply:
Currently FastaSequenceIndex and FastaSequenceIndexCreator are geared towards making an index from a fasta-file, not for generating the index on the fly....so without some serious refactoring that's not going to be easy.